Python/ceval.c

```diff
@@ -5998,3 +6007,38 @@ void Py_LeaveRecursiveCall(void)
 {
     _Py_LeaveRecursiveCall_inline();
 }
+
+static void improve_missing_self_error(PyThreadState* tstate, PyCodeObject* co, Py_ssize_t nargs) {
+
+    if (nargs + 1 != co->co_argcount) {
+        return;
+    }
+    PyObject *type, *val, *tb;
+    _PyErr_Fetch(tstate, &type, &val, &tb);
+
+    if (co->co_argcount > 0 && co->co_varnames && PyTuple_GET_SIZE(co->co_varnames) > 0) {
+        PyObject* first_arg = PyTuple_GET_ITEM(co->co_varnames, 0);
+        int comp = PyUnicode_CompareWithASCIIString(first_arg, "self");
+        if (comp == 0 || comp == -1) {
+            PyErr_Clear();
+            goto exit;
+        }
```

Comment:
`PyUnicode_CompareWithASCIIString()` does not what you think it does. It is like strcmp(). It never raises exception and returns -1 when the first argument is less than the second argument. So your code changes an error message for `def f(x, y)`, but not for `def f(a, b)`.

I suggest to use private functions `_PyUnicode_EqualToASCIIString()` or `_PyUnicode_EqualToASCIIId()` which are more efficient.

Reply:
Thanks for catching that. Fixed!